hydra-node/src/Hydra/HeadLogic.hs

```diff
@@ -691,11 +703,19 @@ onOpenNetworkReqSn env ledger st otherParty sn requestedTxs =
       then continue
       else Wait $ WaitOnSnapshotNumber seenSn
 
+  waitSeenTxs continue =
+    if fromList requestedTxIds `Set.isSubsetOf` fromList (txId <$> seenTxs)
+      then continue
+      else Wait $ WaitOnSeenTxs $ toList $ fromList requestedTxIds `Set.difference` fromList (txId <$> seenTxs)
+
+  resolvedTxs =
+    mapMaybe (\txid -> find (\tx -> txId tx == txid) seenTxs) requestedTxIds
```

Comment:
This is a bit weird, as we _theoretically_ this could drop not-resolved transaction ids.

Could: make this unrepresentable, by combining the `wait` with the resolution and make `waitSeenTxs $ \resolvedTxs ->`

I also combined two statements from the spec in one function [here](https://github.com/input-output-hk/hydra/pull/904/files#diff-f325f10a5757428ff663484ce4cece514a459a08809aa966ecd6d2950f53a67cR674)

Reply:
Not sure I understand your first comment: Dropping unresolvable transactions ids is the point, as we want here to prevent a node to request a snapshot for transactions we have never seen (or possibly have already seen and included in a snapshot).

I will update the continuation to pass the resolved txs, good idea.

Reply:
Yeah I was not clear. Are you familiar with "Parse don't validate?" .. that's the feeling I had here and fail to explain it.

Reply:
I am: https://abailly.github.io/slides/fifty-shades-tdd.html#/parse-dont-validate

Reply:
Interestingly, your comment and the subsequent refactoring I made lead me to a deeper insight: We don't want to `waitForApplicableTxs`, we actually want to `requireApplicableTxs`. I think it's a protocol violation for a leader to send a snapshot with transactions that have been seen (which is what `waitOnSeenTxs` is now doing) but are not applicable to the latest snapshot (`confirmedUTxO`) hence I transformed that into an `Error` instead of a `Wait` outcome.

Happy to discuss that next week as I think I might have missed something.

Reply:
Makes sense that this would be an `Error`